Compare CellAssign annotations using custom marker gene sets from PanglaoDB

allyhawkins opened the pull request:
Closes #225 

This PR looks at cell type annotations of an RMS library using different custom marker gene sets. Previously we used `PanglaoDB` and only considered cells from muscle or connective tissue. However, the submitter annotations indicated that immune cell types (e.g., Monocytes) are also present in this dataset. Here, I tested two different combinations of marker gene sets with `CellAssign`: 

- `combined`: Full set of marker genes from muscle, connective tissue, and all immune cells. 
- `custom`: Marker genes from muscle, connective tissue and only Monoctyes (no other immune cells included). 
(Reading this now I don't know how I came up with those names and if they make any sense at all, so I can rename them if that's confusing) 

A few notes on the analysis in this notebook:

- The notebook requires the predictions from using just the muscle cells. This was done in a previous notebook, so I added the predictions results to this repo so reviewers can run this notebook easily. On that same note, I added the binary matrices and the PanglaoDB reference file to the repo so others could run them. 
- I created a function for turning a marker gene set into a binary gene x cell matrix. 
- I did that primarily because of how long `CellAssign` was taking! I had to add an argument to increase the number of cpus being used and ran the cell assign script with 8 on the server for it to finish within a reasonable time. 
- The main thing I tried to answer here was if adding immune cells to the reference allowed for us to label cell types that weren't previously labeled with just the muscle cells. This is why I specifically looked at Monocytes since that was the population of immune cells in this dataset. I don't think this did a lot because the cells that got labeled Monocytes using `CellAssign` lined up with `Tumor_myoblast` and `Tumor_myocyte`, so we didn't really make the analysis any better? 

Here's a copy of the rendered report: 

[cellassign-combined-markers.nb.html.zip](https://github.com/AlexsLemonade/sc-data-integration/files/12053850/cellassign-combined-markers.nb.html.zip)

sjspielman reviewed (commented):
I'm returning this first with some more overall and high-level comments (also running `renv::restore()` is taking some time over here so I've not run this through yet*!), and then round 2 I'll test out the notebook for real!
I've left some overall comments in the notebook that are mostly small ways to simplify the code a bit, and runtime considerations - let's get some `if` statements in there to not regenerate files that already exist! I also thing an `aws s3 sync` for the processed RDS file itself can't hurt..

> The notebook requires the predictions from using just the muscle cells. This was done in a previous notebook, so I added the predictions results to this repo so reviewers can run this notebook easily. On that same note, I added the binary matrices and the PanglaoDB reference file to the repo so others could run them.

If we have dependent notebooks here, I think it would be worth some more explicit naming like `01-cellassign-blah blah blah` and this notebook with `02-cellassign-blah blah blah`. This should also be reflected in the `README` describing these notebooks.

> Reading this now I don't know how I came up with those names and if they make any sense at all, so I can rename them if that's confusing

They make sense, but could make a bit more sense! It's definitely not something to really "perfect" as it were..Maybe `custom` --> `combined_custom` would be helpful? I've made suggestions throughout to make this happen, but I'm also nervous about breaking the code for such a small thing..Up to you if you want to accept suggestions or ignore them/mark as resolved..
That said, the main thing that's actually confusing me is that you are using `connective tissue` and `muscle` interchangeably, but these don't appear to be separate organs on `PanglaoDB`. Can we pick one term and use throughout? I'd go with `muscle` - no spaces, less typing, and that's how `PanglaoDB` records it. This may involve light textual changes to the other notebook as well.
 
\* footnote from `renv`, I can't seem to install `irlba` on the server due to compilation error which really surprises me since you've presumably installed it, so why not me?!

```
gcc -I"/opt/R/4.2.3/lib/R/include" -DNDEBUG  -I'/home/ccdl/stephanie/sc-data-integration/renv/staging/1/Matrix/include' -I/usr/local/include   -fpic  -g -O2  -c irlb.c -o irlb.o
irlb.c: In function ‘dsdmult’:
irlb.c:590:3: error: too many arguments to function ‘sdmult’
   sdmult (cha, t, one, zero, &chb, &chc, &chol_c);
   ^~~~~~
/opt/R/4.2.3/lib/R/etc/Makeconf:169: recipe for target 'irlb.o' failed
make: *** [irlb.o] Error 1
ERROR: compilation failed for package ‘irlba’
* removing ‘/home/ccdl/stephanie/sc-data-integration/renv/staging/1/irlba’
Error: install of package 'irlba' failed [error code 1]
```
I checked versioning - I have `gcc 7.5.0`, and I would assume you have the same?

allyhawkins commented:
Thanks for the review and helpful comments! 

> I've left some overall comments in the notebook that are mostly small ways to simplify the code a bit, and runtime considerations - let's get some `if` statements in there to not regenerate files that already exist! I also thing an `aws s3 sync` for the processed RDS file itself can't hurt..

I added `if` statements where applicable to generating files and then also the `aws s3 sync` command. I was hesitant to add that here because of the issues we previously had so let me know if you are getting errors with that. 

> If we have dependent notebooks here, I think it would be worth some more explicit naming like `01-cellassign-blah blah blah` and this notebook with `02-cellassign-blah blah blah`. This should also be reflected in the `README` describing these notebooks.

Done!

> They make sense, but could make a bit more sense! It's definitely not something to really "perfect" as it were..Maybe `custom` --> `combined_custom` would be helpful? I've made suggestions throughout to make this happen, but I'm also nervous about breaking the code for such a small thing..Up to you if you want to accept suggestions or ignore them/mark as resolved..

I took this suggestion because I agree it could be clearer! I think I captured all the file renaming necessary for things to not break. 

> That said, the main thing that's actually confusing me is that you are using `connective tissue` and `muscle` interchangeably, but these don't appear to be separate organs on `PanglaoDB`. Can we pick one term and use throughout? I'd go with `muscle` - no spaces, less typing, and that's how `PanglaoDB` records it. This may involve light textual changes to the other notebook as well.

Apologies for the confusion here. RMS is found in both muscle and connective tissue and I tried to capture any organ that would cover that, probably why I was using them both. I changed everything to be muscle in this notebook to be hopefully less confusing! 

> * footnote from `renv`, I can't seem to install `irlba` on the server due to compilation error which really surprises me since you've presumably installed it, so why not me?!

I have not figured this one out yet... I was able to run things on the server while being in the project. I do remember restoring it recently and having issues though. Are you having issues locally or just on the server?

In addition to your comments I made a few other changes based on our discussion yesterday:

- I added labels to the legend for the heatmaps. For some reason `pheatmap` doesn't have an argument to specifically add a legend title, so I changed to using `ComplexHeatmap`.
- I added in comparison of the cell type assignments to cluster assignments. This includes some UMAPs to get some UMAP _vibes_. I will say that I did some annoying wrangling here because I only wanted to plot the top cell types for each reference. Otherwise the UMAPs get very messy. 
- The last addition is heatmaps comparing the cell type assignments for each reference to the clustering assignments. 

Here's the new notebook: 
[03-cellassign-combined-markers.nb.html.zip](https://github.com/AlexsLemonade/sc-data-integration/files/12083815/03-cellassign-combined-markers.nb.html.zip)

sjspielman commented on celltype_annotation/analysis/01-cell-assign-sarcoma.Rmd:
I went to re-run all the renamed files to make sure everything is in order, and I'm getting an error that `excel_celltype_file` doesn't exist.. Is it possible to add `references/Cell_marker_Human.xlsx` to the repo, or is there a reason this needs to be held back?

sjspielman commented on celltype_annotation/references/panglao_combined-custom__mtx.csv:
too many underscores in file name

sjspielman commented on celltype_annotation/analysis/02-cell-assign-sarcoma-marker-genes.Rmd:
Also getting file exists error in this notebook for `'../data/anndata/SCPCL000488_rms_broad_markers.tsv'`

sjspielman commented on celltype_annotation/analysis/03-cellassign-combined-markers.Rmd:
this one is named `-cellassign` but other rmds are `-cell-assign`.. pick your favorite! :) (and update `README` too)

sjspielman reviewed (commented):
What an undertaking with the file names :tada:!! Almost there..
- Getting some path errors for input files used in `01-` and `02-` notebooks . Can those missing files be added to the repo if possible, unless there's a reason we need to hold them back and not disseminate?
- `renv` needs `ComplexHeatmap` - I'm pretty confident the `03-` notebook will run once this package is added (and `__` -> `_` in that one file name)! Happily my `renv` is working today too after yesterday's build extravaganza 🎢

allyhawkins commented:
Okay I _think_ you should be good now. All of the reference files and the prediction matrices from CellAssign have been added. You should be able to run the notebooks now. I also added `ComplexHeatmap` to `renv`. Fingers crossed that your `renv` continues to work today!

sjspielman commented on celltype_annotation/analysis/03-cell-assign-combined-markers.Rmd:
````diff
@@ -0,0 +1,392 @@
+```
+
+```{r}
+# pull out the cell type columns that we want to plot
````
let's add a variable for the `10` below..


```suggestion
top_n_labels <- 10

# pull out the cell type columns that we want to plot
```

sjspielman commented on celltype_annotation/analysis/03-cell-assign-combined-markers.Rmd:
```diff
@@ -0,0 +1,392 @@
+    UMAP_2 = reducedDim(processed_sce, "UMAP")[,2]
+  ) |>
+  # combine all the cell type assignments into one column
+  tidyr::pivot_longer(cols = umap_labels,
```
fewer warnings ^

sjspielman commented on celltype_annotation/analysis/03-cell-assign-combined-markers.Rmd:
```diff
@@ -0,0 +1,392 @@
+                                         clusters,
+                                         cluster_cols = FALSE,
+                                         cluster_rows = TRUE,
+                                         title = label))
```
```suggestion
                                         title = glue::glue("{label} across clusters")))
```

sjspielman commented on celltype_annotation/analysis/03-cell-assign-combined-markers.Rmd:
```diff
@@ -0,0 +1,392 @@
+One thing to note is that when using `SingleR` we are able to assign cells to `muscle cells` and `skeletal muscle cells`. 
+Although the assignments from `SingleR` (see `singler-rms-comparison.Rmd`) don't completely line up with the submitter annotations, identifying cells as muscle cells rather than labeling most of them as "other" does seem more informative to me. 
+- The more marker genes you have the more time `CellAssign` takes. 
+We already knew this, but this was a new record. 
```
🚀

sjspielman commented on celltype_annotation/analysis/03-cell-assign-combined-markers.Rmd:
````diff
@@ -0,0 +1,392 @@
+```{r, fig.height=5, fig.width=10}
+# heatmap of prediction scores 
+all_predictions |>
+  purrr::imap(\(prediction_mtx, organ_type){
````
Something is funky here.. the last heatmap is overlayed with a second `combined_custom` heatmap. I played around with this for a bit, and the last matrix is always overlayed with the previous one. Changing to `iwalk` seems to do the trick!

```suggestion
  purrr::iwalk(\(prediction_mtx, organ_type){
```


<img width="568" alt="Screenshot 2023-07-19 at 11 20 02 AM" src="https://github.com/AlexsLemonade/sc-data-integration/assets/4701111/189618a8-c243-4961-9329-ee47a3ab2015">

sjspielman reviewed (approved):
A couple small things to tweak for clarity and reduce silly `purrr::map()` side effects, but otherwise looks great! Thanks for making all the changes along the way!

I don't feel that I need to see this again unless you want me too, so as long as everything looks ok with the rendered  HTML after incorporating suggestions, go for it 🚀
